aws-sdk-core/lib/aws-sdk-core/shared_credentials.rb

```diff
@@ -67,15 +67,35 @@ def default_path
 
     def load_from_path
       profile = load_profile
-      @credentials = Credentials.new(
+      credentials = Credentials.new(
         profile['aws_access_key_id'],
         profile['aws_secret_access_key'],
         profile['aws_session_token']
       )
+      @credentials = if role_arn = profile['role_arn']
+        AssumeRoleCredentials.new(
+          role_session_name: profile['role_session_name'] || "AWS-Ruby-session-%s"%[Time.now.to_i],
+          external_id: profile['external_id'],
+          role_arn: role_arn,
+          credentials: credentials
+        ).credentials
+      else
+        credentials
+      end
     end
 
     def load_profile
       if profile = profiles[profile_name]
+        if profile.key?('role_arn')
+          if source_profile = profile['source_profile']
+            profile['aws_access_key_id'] = profiles[source_profile]['aws_access_key_id']
```

Comment:
Yeah, I agreed it would be more convenient to have access_key and session_token in the non-source profile. But I made these changes so that it can be uniform with python and java implementations. In botocore, for example, source profile is a mandatory field, and access_key and session_token are only retrieved from source profile. Please refer https://github.com/boto/botocore/blob/develop/botocore/credentials.py#L840-L866 for more information. I believe it is the same idea in java implementation.